src/thor/route_action.cc

```diff
@@ -489,50 +513,87 @@ void thor_worker_t::path_arrive_by(Api& api, const std::string& costing) {
     // if we just made a leg that means we are done recording which algorithms were used
     if (path.empty())
       algorithms.clear();
-  }
 
+    return true;
+  };
+
+  auto correlated = options.locations();
+  bool allow_retry = true;
+
+  // For each pair of locations
+  auto origin = ++correlated.rbegin();
+  while (origin != correlated.rend()) {
+    auto destination = std::prev(origin);
+    if (!route_two_points(origin, destination)) {
+      // if routing failed because an intermediate waypoint was snapped to the low reachability road
+      // (such road lies in a small connectivity component that is not connected to other locations)
+      // we should leave only high reachability candidates and try to route again
+      if (allow_retry && is_through_point(*destination) &&
+          !is_high_reachable(*destination, destination->path_edges(0))) {
+        allow_retry = false;
+        // for each intermediate waypoint remove candidates with low reachability
+        correlated = options.locations();
+        for (auto loc = std::next(correlated.begin()); loc != std::prev(correlated.end()); ++loc) {
+          remove_edges_with_low_reachability(*loc);
+          // it doesn't make sense to continue if there are no more path edges
+          if (loc->path_edges_size() == 0)
+            // no route found
+            throw valhalla_exception_t{442};
+        }
+        // reset state and try to route again
+        route = nullptr;
+        first_edge = {};
+        vias.clear();
+        path.clear();
+        algorithms.clear();
+        trip.mutable_routes()->Clear();
+        origin = ++correlated.rbegin();
+        continue;
```

Comment:
resets the entire state of all the legs of the route and starts completely over from the beginning doing all the legs over.

Reply:
is it proposition for comment ? to make more detailed ?

Reply:
It's just me clarifying my understanding or anyone else's who is following along haha